### PR TITLE
Add Skeleton Templates

### PR DESCRIPTION
**Pull request by aaprasad**
### Description
Finished MVP for skeleton templates:
- stored skeleton jsons corresponding to each of the [reference datasets](https://sleap.ai/datasets.html) in `sleap/skeletons`
- added dropdown to select from one of the above skeleton templates
- Modified the skeleton's load button to load from the selected skeleton or a user defined file labeled `Custom` in the dropdown 

### Types of changes

- [ ] Bugfix
- [X] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
#1106 

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

File: tests/gui/test_commands.py [hunk 1]
```diff
@@ -391,11 +392,18 @@ def assert_skeletons_match(new_skeleton: Skeleton, skeleton: Skeleton):
...
+        )  # Original function uses `QComboBox.currentText()`
+        if template == "Custom":
+            # Original function opens FileDialog here
+            print("Loading from file dialog")
+            filename = params["filename_in"]
+        else:
+            print("Loading from template")
+            filename = get_package_file(f"sleap/skeletons/{template}.json")
         if len(filename) == 0:
+            print("Whoops, filename 0.")
```
Comment (roomrys) [inline tests/gui/test_commands.py:406]: Remove print statements
```suggestion
            filename = params["filename_in"]
        else:
            filename = get_package_file(f"sleap/skeletons/{template}.json")
        if len(filename) == 0:
```

File: sleap/gui/commands.py [hunk 2]
```diff
@@ -1863,15 +1864,24 @@ def delete_extra_skeletons(labels: Labels):
...
+            filename, selected_filter = FileDialog.open(
+                context.app,
+                dir=None,
+                caption="Open skeleton...",
+                filter=";;".join(filters),
+            )
+        else:
+            filename = get_package_file(
+                f"sleap/skeletons/{template}.json"
+            )  # load from selected preset
```
Comment (roomrys) [inline sleap/gui/commands.py:1881]: Nitpick: [Google style](https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments) comments and punctuation
```suggestion
        # Get selected template from dropdown
        template =  context.app.skeletonTemplates.currentText()
        
        # Check whether to load from file or preset
        if template == "Custom": 
            filename, selected_filter = FileDialog.open(
                context.app,
                dir=None,
                caption="Open skeleton...",
                filter=";;".join(filters),
            )
        else:  # Load from a selected preset
            filename = get_package_file(f"sleap/skeletons/{template}.json")
```

**Review by roomrys — CHANGES_REQUESTED** [review]
Few small non-functional changes. Otherwise LGTM. Good to start next PR after small changes.

**Comment by talmo** [comment]
Notes on finishing the integration:
- Preview image + description should show on selecting the skeleton from the dropdown, not just after clicking Load
- Description should have auto-generated text describing the skeleton, e.g.: `f"Nodes: {len(skeleton.nodes)}. Edges: {len(skeleton.edges)}. Node names: {'\n'.join(skeleton.node_names)}"`
- We need a "Clear" button that deletes the current skeleton so that it's easier to actually _switch_ between skeletons instead of merging
    - **Danger zone:** Clearing the skeleton might delete all annotations from existing instances? At a minimum we should add a confirmation dialog box.
- Unify node names so that the same type of body part is named the same across skeleton variants (e.g., "tail_base" and "tti" in `mice_of` vs `mice_hc`)

File: sleap/gui/app.py [hunk 3]
```diff
@@ -1022,7 +1119,8 @@ def _add_button(to, label, action, key=None):
         hbw.setLayout(hb)
         vb.addWidget(hbw)
         gb.setLayout(vb)
-        skeleton_layout.addWidget(gb)
+        graph_tabs.addTab(gb, "Nodes")
+        # skeleton_layout.addWidget(graph_tabs)
```
Comment (roomrys) [inline sleap/gui/app.py:1123]: Remove

File: sleap/gui/app.py [hunk 4]
```diff
@@ -1003,9 +1008,101 @@ def _add_button(to, label, action, key=None):
...
+
+        hb = QHBoxLayout()
+        _add_button(hb, "Load Skeleton", self.commands.openSkeleton)
+        _add_button(hb, "Save Skeleton", self.commands.saveSkeleton)
+
+        hbw = QWidget()
+        hbw.setLayout(hb)
+        skeleton_layout.addWidget(hbw)
 
         gb = QGroupBox("Nodes")
```
Comment (roomrys) [inline sleap/gui/app.py:1103]: We no longer need this to be in a `GroupBox` since it is contained in its own tab (same for the "Edges" `GroupBox`)

File: sleap/util.py [hunk 5]
```diff
@@ -21,9 +21,13 @@
...
 import rapidjson
 import yaml
+import base64
 
 from typing import Any, Dict, Hashable, Iterable, List, Optional
 
+from PIL import Image
+from io import BytesIO
+
 import sleap.version as sleap_version
```
Comment (roomrys) [inline sleap/util.py:31]: [Reorganize these imports](https://realpython.com/python-import/#imports-style-guide)

File: tests/nn/test_viz.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+from sleap.instance import LabeledFrame, Track
+from sleap.io.dataset import Labels
+from sleap.nn.viz import generate_skeleton_preview_image
+
+
+def test_generate_skeleton_preview_image(
+    centered_pair_predictions_slp_path: str,
+    centered_pair_vid_path: str,
+):
+    """Encode/decode preview images for all skeletons in sleap.skeletons directory."""
```
Comment (roomrys) [inline tests/nn/test_viz.py:13]: Just encode (no decode)

File: tests/gui/test_dialogs.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    skeleton_new = flies13_skeleton
+    rename_nodes, delete_nodes, add_nodes = OpenSkeleton.compare_skeletons(
+        skeleton, skeleton_new
+    )
+
+    win = ReplaceSkeletonTableDialog(
+        rename_nodes,
+        delete_nodes,
+        add_nodes,
+    )
```
Comment (roomrys) [inline tests/gui/test_dialogs.py:50]: Also check case when `win.table == None` (no add/delete nodes).

**Review by roomrys — CHANGES_REQUESTED** [review]

**Comment by roomrys** [comment]
> Notes on finishing the integration:
> 
> 1. Preview image + description should show on selecting the skeleton from the dropdown, not just after clicking Load
> 2. Description should have auto-generated text describing the skeleton, e.g.: `f"Nodes: {len(skeleton.nodes)}. Edges: {len(skeleton.edges)}. Node names: {'\n'.join(skeleton.node_names)}"`
> 3. We need a "Clear" button that deletes the current skeleton so that it's easier to actually _switch_ between skeletons instead of merging
>   
>   * **Danger zone:** Clearing the skeleton might delete all annotations from existing instances? At a minimum we should add a confirmation dialog box.
> 4. Unify node names so that the same type of body part is named the same across skeleton variants (e.g., "tail_base" and "tti" in `mice_of` vs `mice_hc`)

All of the above except for 3 have been implemented. I think the replace skeleton table takes care of this though?

File: sleap/io/dataset.py [hunk 8]
```diff
@@ -2618,6 +2618,7 @@ def video_callback(
                         for i, filename in enumerate(filenames):
                             if missing[i]:
                                 filenames[i] = new_paths[i]
+                                missing[i] = False
```
Comment (roomrys) [inline sleap/io/dataset.py:2621]: Flagging this since it is outside of the scope of the PR (noticed that when loading file through API, found videos are not replaced if USE_DUMM_FOR_MISSING_VIDEOS was set to True) - this line fixes that and only uses dummy for truly missing videos

File: sleap/skeleton.py [hunk 9]
```diff
@@ -92,11 +92,16 @@ class Skeleton:
     The skeleton represents the constituent parts of the animal whose
     pose is being estimated.
 
-    An index variable used to give skeletons a default name that should
-    be unique across all skeletons.
+    Attributes:
+        _skeleton_idx: An index variable used to give skeletons a default name that should
+        be unique across all skeletons.
```
Comment (roomrys) [inline sleap/skeleton.py:97]: Nit: Add tab here

File: sleap/nn/viz.py [hunk 10]
```diff
@@ -296,3 +300,87 @@ def plot_bbox(bbox, **kwargs):
...
+        square_bb: A boolean flag for whether or not the preview image should be a square image
+        thumbnail_size: A tuple of (w,h) for what the size of the thumbnail image should be
+
+    Returns:
+        A byte string encoding of the preview image.
+    """
+
+    def get_square_bounding_box(bb):
+        """
+        Convert rectangular bounding box to square bounding box
```
Comment (roomrys) [inline sleap/nn/viz.py:321]: Nit: this should be one line

**Review by roomrys — COMMENTED** [review]
Nits

**Review by roomrys — APPROVED** [review]

**Comment by roomrys** [comment]
The new skeleton dialog:
---

Open
---
image

Collapsed
---
image

**Review by roomrys — APPROVED** [review]

**Review by talmo — APPROVED** [review]
Let's take a leap of faith and see what happens!!
